vyper/parser/function_definitions/parse_public_function.py

```diff
@@ -138,12 +135,29 @@ def parse_public_function(code: ast.FunctionDef,
             mem_pos, _ = context.memory_allocator.increase_memory(32 * get_size_of_type(arg.typ))
             context.vars[arg.name] = VariableRecord(arg.name, mem_pos, arg.typ, False)
         else:
-            context.vars[arg.name] = VariableRecord(
-                arg.name,
-                MemoryPositions.RESERVED_MEMORY + arg.pos,
-                arg.typ,
-                False,
-            )
+            if sig.name == '__init__':
+                context.vars[arg.name] = VariableRecord(
+                    arg.name,
+                    MemoryPositions.RESERVED_MEMORY + arg.pos,
+                    arg.typ,
+                    False,
+                )
+            elif i >= default_args_start_pos:  # default args need to be allocated in memory.
+                default_arg_pos, _ = context.memory_allocator.increase_memory(32)
```

Comment:
Does this work for complex types - structs or arrays?

Reply:
Yes I believe so :) only bytearays are a problem because of dynamic nature and existing mem copies.